src/transaction/log_page_buffer.c

```diff
@@ -3361,6 +3364,250 @@ logpb_write_toflush_pages_to_archive (THREAD_ENTRY * thread_p)
+ *       │           │ Remove the EOF at the append_lsa on the memory buffer.
+ *       └───────────┘ This will be flushed in the next time.
+ *
+ * (*): It's done to append log pages containing an incomplete log record without flushing in the middle.
+ * 
+ *      See logpb_catchup_end_append() and logpb_flush_all_append_pages() as well for the complete picture.
+ *     
+ *   log_pgptr(in): a log page pointer to append
+ */
+static void
+logpb_catchup_start_append (THREAD_ENTRY * thread_p)
+{
+  assert (LOG_CS_OWN_WRITE_MODE (thread_p));
+
+  switch (log_Pb.partial_append.status)
+    {
+    case LOGPB_APPENDREC_SUCCESS:
+      log_Pb.partial_append.status = LOGPB_APPENDREC_IN_PROGRESS;
+      break;
+    case LOGPB_APPENDREC_IN_PROGRESS:
+    case LOGPB_APPENDREC_PARTIAL_FLUSHED_END_OF_LOG:
+      /* keep the status */
+      break;
+    default:
+      /* invalid state */
+      assert_release (false);
+    }
+}
+
+/*
+ * logpb_catchup_end_append - Finish appending a page during catch-up.
+ *      If the appended page has only parital record in it, keep partial appending status.
+ *      Otherwise, complete appending pages, in the partial appending case, or a page.
+ *        - Set the append_lsa and prev_lsa to valid ones.
+ *        - Remove the temporary EOF If log pages are flushed halfway through.
+ *        - Set the partial appending status to LOGPB_APPENDREC_SUCCESS.
+ * 
+ *    See logpb_catchup_start_append() for details of the status transition.
+ *
+ *   log_pgptr(in): a log page pointer appended
+ */
+static void
+logpb_catchup_end_append (THREAD_ENTRY * thread_p, const LOG_PAGE * const log_pgptr)
+{
+  assert (LOG_CS_OWN_WRITE_MODE (thread_p));
+
+  const bool have_only_partial_record_in_page = log_pgptr->hdr.offset == NULL_LOG_OFFSET;
+
+  if (have_only_partial_record_in_page)
+    {
+      // Do nothing. It will be finished up at a following call.
+      return;
+    }
+
+  assert (log_Gl.hdr.append_lsa.pageid == log_pgptr->hdr.logical_pageid);
+  log_Gl.hdr.append_lsa.offset = log_pgptr->hdr.offset;
+
+  LOG_RECORD_HEADER *log_rec = LOG_GET_LOG_RECORD_HEADER (log_pgptr, &log_Gl.hdr.append_lsa);
+  log_Gl.append.prev_lsa = log_rec->back_lsa;
+
+  switch (log_Pb.partial_append.status)
+    {
+    case LOGPB_APPENDREC_IN_PROGRESS:
+      /* success, fall through */
+      break;
+    case LOGPB_APPENDREC_PARTIAL_FLUSHED_END_OF_LOG:
+      {
+	LOG_RECORD_HEADER *origin_append_log_record = LOG_GET_LOG_RECORD_HEADER (log_pgptr, &log_Gl.hdr.append_lsa);
+
+	// The temporary EOF log record at the prev_lsa has to be removed.
+	log_Pb.partial_append.status = LOGPB_APPENDREC_PARTIAL_ENDED;
+	logpb_flush_all_append_pages (thread_p);
+	assert (log_Pb.partial_append.status == LOGPB_APPENDREC_PARTIAL_FLUSHED_ORIGINAL);
+
+	// A new EOF has been appended at the append_lsa in logpb_flush_all_append_pages. Remove it as well.
+	LOG_RECORD_HEADER *append_log_rec = LOG_GET_LOG_RECORD_HEADER (log_Gl.append.log_pgptr, &log_Gl.hdr.append_lsa);
```

Comment:
memcpy (log_Gl.append.log_pgptr, log_pgptr, LOG_PAGESIZE) is executed before calling this function. So, I think that log_Gl.append.log_pgptr has same data with log_pgptr.

At 3516 line, origin_append_log_record is copied into append_log_rec. This seems like a meaningness behavior.

In my opinion, it would be good to first conduct an off-line review focusing on the partial append mechanism.

Reply:
In `logpb_flush_all_append_pages` with the status of `LOGPB_APPENDREC_PARTIAL_ENDED`, an EOF is appended at the `log_Gl.hdr.append_lsa`. It is mainly to remove the previous EOF temporarily appended at the `prev_lsa`, but we always need `EOF`, so it's then at the `append_lsa`. We should remove it in the memory buffer, and it will overwrite the EOF when it's flushed.